gosh/framework.sh.tpl

```diff
@@ -0,0 +1,114 @@
+#!/bin/sh
+
+set -u
+
+GOSH_FAILED=0
+
+if command -v gt >/dev/null 2>&1; then
+    readonly GT_BIN="gt"
+else
+    echo "Error: gt command not found in PATH" >&2
+    exit 1
+fi
+
+# Send a signal to the test runner
+# Args:
+#   $1 - The operation to perform (e.g., start, stop, status)
+#   $2 - The name of the test
+#   $3 - The exit code of the test (optional, default is 0)
+#   $4 - An optional message to include (default is "")
+_gosh_signal() {
+    local op="${1}"
+    local test_name="${2}"
+    local exit_code="${3:-0}"
+    local message="${4:-}"
+
+    "${GT_BIN}" ctrl \
+        --op="$op" \
+        --test-name="$test_name" \
+        --exit-code="$exit_code" \
+        --addr="${GOSH_CONTROL_ADDR}" \
+        --message="${message}"
+}
+
+# Check if a function exists
+_gosh_function_exists() {
+    local output
+    output=$(type "$1" 2>/dev/null) || return 1
+    case "$output" in
+        *'function'*)
+            return 0
+            ;;
+        *)
+            return 1
+            ;;
+    esac
+}
+
+# Run a wrapped test with proper pipe redirection. The wrapped test function
+# blocks until the test completes and the runner acks the "end" signal.
+# Args:
+#   $1 - The name of the test
+#   $2 - The stdout pipe path
+#   $3 - The stderr pipe path
+# Returns:
+#   The exit code of the test
+_gosh_run_test() {
+    local test_name="${1}"
+    local stdout_pipe="${2}"
+    local stderr_pipe="${3}"
+
+    _gosh_signal "start" "${test_name}"
+
+    # User defined test(s) run in this subshell
+    (
+      set -eou pipefail
+
+      trap '_gosh_teardown' EXIT
+
+      _gosh_teardown() {
+          if _gosh_function_exists "teardown"; then
+              teardown > "${stdout_pipe}" 2> "${stderr_pipe}"
+          fi
+      }
+
+      if _gosh_function_exists "setup"; then
+          setup
+      fi
+
+      "${test_name}"
+    ) < /dev/null > "${stdout_pipe}" 2> "${stderr_pipe}"
+
+    local test_exit_code=$?
+
+    # Signal the test completion with the exit code
+    _gosh_signal "end" "${test_name}" "${test_exit_code}"
+
+    return "${test_exit_code}"
+}
+
+# Inline the newly rendered user script content, the shbang will be preserved, but will always be treated as a comment by the actual interpreter above
+{{loadScript}}
+
+# Run the one-time setup if defined
+if _gosh_function_exists "setup_once"; then
+    setup_once
+fi
```

Comment:
I think if we just define `setup_once` above as:
```
setup_once() { return 0; }
```

then we can ditch `_gosh_function_exists` and just call it.  if the `{{loadScript}}` redefines it then its version will get used.

Reply:
genius!